enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,419 @@
+---
+title: out-of-tree-provider-support
+authors:
+  - "@Danil-Grigorev"
+  - "@mfedosin"
+reviewers:
+  - "@enxebre"
+  - "@JoelSpeed"
+  - "@crawford"
+  - "@derekwaynecarr"
+  - "@eparis"
+  - "@mrunalp"
+  - "@sttts"
+approvers:
+  - "@enxebre"
+  - "@JoelSpeed"
+  - "@crawford"
+  - "@derekwaynecarr"
+  - "@enxebre"
+  - "@eparis"
+  - "@mrunalp"
+  - "@sttts"
+creation-date: 2020-08-31
+last-updated: 2020-09-29
+status: implementable
+see-also:
+  - "/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md"  
+replaces:
+superseded-by:
+---
+
+# Add support for out-of-tree cloud provider integration
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposal describes steps needed to add support for out-of-tree providers in cloud providers currently supported by Openshift. This includes: AWS, Azure, GCP, vSphere, Openstack.
+
+As a pioneer platform, it is proposed to use OpenStack. 
+
+OpenStack external cloud provider fixes many issues and limitations with the in-tree cloud provider, such as it's reliance on Nova metadata service. For OpenStack platform, this means the possibility for deploying on provider networks and at the edge.
+
+## Motivation
+
+The Kubernetes community is currently moving towards exclusion of the in-tree cloud-provider code from the core Kubernetes repository, and instead introducing out-of-tree provider support, making them independent from the core Kubernetes release cycle, and simplifying the process of development and e2e testing for those plugins. Previously each cloud-provider was located in the [`legacy-cloud-providers`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/legacy-cloud-providers) and now the support and main development is done in the dedicated cloud-provider [repositories](https://github.com/kubernetes?q=cloud-provider-). They plan to remove the in-tree support completely in the future releases, approximetly in [v1.21](https://kubernetes.slack.com/archives/CPPU7NY8Y/p1589289501020600).
+
+### Goals
+
+- Prepare openshift components to accomodate the out-of-tree plugins and remove support for the in-tree implementation.
+- Describe an approach how an `external` cloud provider configuration for the cloud will be supported.
+- Provide the means to allow generic `CCM` component registration via the provider specific upstream implemetation.
+- Concider support for running switching cloud configurations in a single cluster.
```

Comment:
Is this as in, switching from GCP to AWS? If so, I don't think that should be a goal of this project

Or is this switching from in-tree to out-of-tree as an option? Also don't think that should be a goal here

Minor typo btw
```suggestion
- Consider support for running switching cloud configurations in a single cluster.
```